spatial_cell_kernels.hpp

```diff
@@ -0,0 +1,682 @@
+   // const vmesh::LocalID nBlocksAfterAdjust = nBlocksBeforeAdjust + n_with_replace_new - n_to_delete;
+   // const vmesh::LocalID nBlocksToChange = n_with_replace_new + n_with_replace_old + n_to_delete;
+
+   const vmesh::LocalID nToAdd = list_with_replace_new->size();
+   const vmesh::LocalID nToRemove = list_delete->size() + list_to_replace->size();
+   const vmesh::LocalID nBlocksAfterAdjust = nBlocksBeforeAdjust + nToAdd - nToRemove;
+   const vmesh::LocalID nBlocksToChange = nToAdd > nToRemove ? nToAdd : nToRemove;
+
+   gpu_rhoLossAdjust[0] = 0.0;
+   returnLID[0] = nBlocksBeforeAdjust;
+   returnLID[1] = nBlocksAfterAdjust;
+   returnLID[2] = nBlocksToChange;
+   // Should we grow the size?
+   if (nBlocksAfterAdjust > nBlocksBeforeAdjust) {
+      if ((nBlocksAfterAdjust <= vmesh->capacity()) && (nBlocksAfterAdjust <= blockContainer->capacity())) {
+         returnLID[3] = 1; // Resize on-device will work.
+         vmesh->device_setNewSize(nBlocksAfterAdjust);
+         blockContainer->setNewSize(nBlocksAfterAdjust);
+      } else {
+         returnLID[3] = 0; // Need to recapacitate and resize from host
+      }
+   } else {
+      // No error as no resize.
+      returnLID[3] = 1;
+   }
+}
+
+/** Mini-kernel for adjusting vmesh and VBC size on-device aftewards (shrink only) */
+__global__ void resize_vbc_kernel_post(
+   vmesh::VelocityMesh *vmesh,
+   vmesh::VelocityBlockContainer *blockContainer,
+   vmesh::LocalID nBlocksAfterAdjust
+   ) {
+   vmesh->device_setNewSize(nBlocksAfterAdjust);
+   blockContainer->setNewSize(nBlocksAfterAdjust);
+}
+
+/** GPU kernel for updating blocks based on generated lists */
+__global__ void __launch_bounds__(WID3,4) update_velocity_blocks_kernel(
+   vmesh::VelocityMesh *vmesh,
+   vmesh::VelocityBlockContainer *blockContainer,
+   split::SplitVector<vmesh::GlobalID>* list_with_replace_new,
+   split::SplitVector<Hashinator::hash_pair<vmesh::GlobalID,vmesh::LocalID>>* list_delete,
+   split::SplitVector<Hashinator::hash_pair<vmesh::GlobalID,vmesh::LocalID>>* list_to_replace,
+   split::SplitVector<Hashinator::hash_pair<vmesh::GlobalID,vmesh::LocalID>>* list_with_replace_old,
+   vmesh::LocalID nBlocksBeforeAdjust,
+   vmesh::LocalID nBlocksToChange,
+   vmesh::LocalID nBlocksAfterAdjust,
+   Realf* gpu_rhoLossAdjust
+   ) {
+   //const int gpuBlocks = gridDim.x;
+   //const int blocki = blockIdx.x;
+   //const int blockSize = blockDim.x; // WID3
+   const uint ti = threadIdx.x; // [0,blockSize)
+
+   // Each GPU block / workunit could manage several Vlasiator velocity blocks at once.
+   // However, thread syncs inside the kernel prevent this.
+   //const uint vlasiBlocksPerWorkUnit = WARPSPERBLOCK * GPUTHREADS / WID3;
+   //const uint workUnitIndex = ti / WID3; // [0,vlasiBlocksPerWorkUnit)
+   //const uint index = blocki * vlasiBlocksPerWorkUnit + workUnitIndex; // [0,nBlocksToChange)
+   //const uint vlasiBlocksPerWorkUnit = 1;
+   //const uint workUnitIndex = 1;
```

Comment:
drop this?

Reply:
This is a potential future TODO still, I think, so stays.